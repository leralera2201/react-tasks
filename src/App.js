import React from 'react';
import { Switch, Route } from 'react-router-dom';

import Header from "@components/Header";

import Buttons from "./pages/Buttons";
import Typography from "./pages/Typography";
import Dropdown from "./pages/Dropdown";
import Select from "./pages/Select";
import Alert from "./pages/Alert";


function App() {
  return (
    <div className="App">
        <Header />
        <Switch>
            <Route path="/" exact component={Buttons}/>
            <Route path="/typography" exact component={Typography}/>
            <Route path="/dropdown" exact component={Dropdown}/>
            <Route path="/select" exact component={Select}/>
            <Route path="/alert" exact component={Alert}/>
        </Switch>
    </div>
  );
}

export default App;
