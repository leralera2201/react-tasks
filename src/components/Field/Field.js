import React from "react";
import { FormGroup, Input, Label } from "reactstrap";

const Field = ({ label, name, ...otherProps }) => {
  return (
    <FormGroup>
      <Label htmlFor={name}>{label}</Label>
      <Input id={name} placeholder={label} name={name} {...otherProps} />
    </FormGroup>
  );
};

export default Field;
