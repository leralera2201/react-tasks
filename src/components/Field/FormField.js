import React from 'react';
import { FormGroup, Input, Label } from 'reactstrap';
import { Field } from 'react-final-form';
import PropTypes from 'prop-types';
import Errors from "../Errors";

const FormField = ({ label, name, type }) => {
  return (
    <Field name={name}>
      {({ input, meta: {
          invalid, active, error, submitFailed, visited
      }}) => (
        <FormGroup>
          <Label htmlFor={name}>{label}</Label>
          <Input
            {...input}
            id={name}
            placeholder={label}
            name={name}
            type={type}
            invalid={invalid && !active && (submitFailed || visited)}
          />
          {invalid && !active && (submitFailed || visited) &&
              <Errors errors={Object.values(error)} />
          }
        </FormGroup>
      )}
    </Field>
  );
}

FormField.propTypes = {
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
};

export default FormField;
