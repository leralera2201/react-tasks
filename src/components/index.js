export { default as Header } from './Header';
export { default as Footer } from './Footer';
export { default as Error } from './Error';
export { default as Loader } from './Loader';
export { default as BookCard } from './BookCard';
export { default as BooksPagination } from './BooksPagination';
