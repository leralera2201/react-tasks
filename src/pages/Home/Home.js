import React from 'react';
import { Container, Jumbotron } from 'reactstrap';

const Home = () => (
  <div className="container pt-5">
    <Jumbotron fluid>
      <Container fluid>
        <h1 className="display-3">Fluid jumbotron</h1>
        <p className="lead">
          This is a modified jumbotron that occupies the entire horizontal
          space of its parent.
        </p>
      </Container>
    </Jumbotron>
  </div>
);

export default Home;
