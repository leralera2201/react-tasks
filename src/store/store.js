import { createStore, applyMiddleware, compose } from "redux";
import createSagaMiddleware from "redux-saga";

import rootReducer from "./reducers";
import allSagas from "./sagas";

const sagaMiddleware = createSagaMiddleware();
const composeEnhancers =
  (typeof window !== "undefined" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(allSagas);

export default store;
