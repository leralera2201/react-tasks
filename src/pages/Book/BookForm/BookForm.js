import React, { Component } from "react";
import { Button, Form } from "reactstrap";
import { connect } from "react-redux";

import Field from "@components/Field";

import { createBookStart, updateBookStart } from "../actions/book.actions";

const fields = {
  title: {
    label: "Title",
    type: "text",
    value: "",
  },
  description: {
    label: "Description",
    type: "text",
    value: "",
  },
  excerpt: {
    label: "Excerpt",
    type: "text",
    value: "",
  },
  publishDate: {
    label: "Publish Date",
    type: "date",
    value: "",
  },
  pageCount: {
    label: "Page Count",
    type: "number",
    value: "",
  },
};

export class BookForm extends Component {
  state = {
    ...fields,
  };

  componentDidMount() {
    const {
      location: { state },
    } = this.props;
    if (state && state.book) {
      Object.entries(state.book).forEach(([fieldName, value]) => {
        this.setState({
          [fieldName]: {
            ...this.state[fieldName],
            value,
          },
        });
      });
    }
  }

  changeHandler = (event) => {
    const { name, value } = event.target;
    const data = this.state;
    data[name] = { ...data[name], value };

    this.setState({ ...data });
  };

  submitHandler = () => {
    const {
      createBook,
      updateBook,
      history,
      match: {
        params: { id },
      },
    } = this.props;
    const data = Object.entries(this.state).reduce((result, field) => {
      const [fieldName, { value }] = field;
      result[fieldName] = value;
      return result;
    }, {});
    if (id) {
      updateBook(data);
    } else {
      createBook(data);
    }
    this.setState({ ...fields });
    history.push({
      pathname: "/books",
      state: { withoutFetch: true },
    });
  };

  render() {
    const data = this.state;
    const {
      match: {
        params: { id },
      },
    } = this.props;
    return (
      <div className="container pt-5">
        <Form>
          {Object.entries(data).map(([fieldName, field]) => (
            <Field
              key={fieldName}
              type={field.type}
              label={field.label}
              name={fieldName}
              value={field.value}
              onChange={this.changeHandler}
            />
          ))}
          <Button onClick={this.submitHandler}>{id ? "Update" : "Add"}</Button>
        </Form>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  createBook: (book) => dispatch(createBookStart(book)),
  updateBook: (book) => dispatch(updateBookStart(book)),
});

export default connect(null, mapDispatchToProps)(BookForm);
