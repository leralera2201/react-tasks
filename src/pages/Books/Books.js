import React, { Component } from "react";
import { connect } from "react-redux";

import { Loader, BookCard, BooksPagination, Error } from "../../components";

import {fetchBooksStart, paginateBooks} from "./actions/books.actions";
import {
  booksFetchErrorSelector,
  booksFetchIsStatusInProgress,
  booksFetchItemsSelector,
  booksFetchCurrentPageSelector,
  booksFetchBooksPerPagePageSelector,
  booksFetchCurrentItemsSelector
} from "./selectors/books.selectors";
import {Button} from "reactstrap";

class Books extends Component {
  componentDidMount() {
    const { getAllBooks, paginate, location: {state} } = this.props;
    if(state && state.withoutFetch) {
      paginate()
      return
    }
    getAllBooks();
  }

  createBook = () => {
    const { history } = this.props;
    history.push('/books/create')
  }

  render() {
    const {
      loading,
      currentBooks,
      error,
      booksPerPage,
      currentPage,
      books,
      paginate,
    } = this.props;
    const booksLength = books.length;

    const pages = [];

    const totalPages = Math.ceil(booksLength / booksPerPage);
    for (let i = 1; i <= totalPages; i++) {
      pages.push(i);
    }

    return (
      <div className="pt-5">
        <div className="container">
          <Button onClick={this.createBook}>
            Create book
          </Button>
        {loading && <Loader />}
        {!loading && !error && currentBooks.length > 0 && (
            <>
              <div className="flex-space-between">
                {currentBooks.map((book) => (
                  <BookCard book={book} key={book.id} />
                ))}
              </div>
              <BooksPagination
                currentPage={currentPage}
                onPaginate={paginate}
                pages={pages}
              />
            </>
        )}
        {!error && !loading && booksLength === 0 && <div>No books</div>}
        {error && <Error error={error} />}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  loading: booksFetchIsStatusInProgress(state),
  books: booksFetchItemsSelector(state),
  error: booksFetchErrorSelector(state),
  currentPage: booksFetchCurrentPageSelector(state),
  booksPerPage: booksFetchBooksPerPagePageSelector(state),
  currentBooks: booksFetchCurrentItemsSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  getAllBooks: () => dispatch(fetchBooksStart()),
  paginate: (page) => dispatch(paginateBooks(page)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Books);
