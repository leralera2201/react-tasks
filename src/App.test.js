import React from 'react';
import {mount, shallow} from "enzyme";
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import App from "./App";
import {Footer, Header} from "./components";

import { BrowserRouter as Router } from 'react-router-dom';
import Books from "./pages/Books";
import BookForm from "./pages/Book/BookForm";
import Field from "./components/Field";
import store from "./store/store";
import Home from "./pages/Home";

const mockStore = configureMockStore();
// const store = mockStore({});

let app = <Router><App /></Router>

const formRouterWithProvider = (route, component = <App/>) => <MemoryRouter initialEntries={[route]}>
    <Provider store={store}>
        {component}
    </Provider>
</MemoryRouter>


describe('App', () => {
    it('renders the Header', () => {
        const wrapper = mount(app);
        expect(wrapper.find(Header).length).toEqual(1);
    });

    it('renders the Home', () => {
        const wrapper = mount(formRouterWithProvider('/'));
        expect(wrapper.find(Home).length).toEqual(1);
    });

    it('renders the Books', () => {
        const wrapper = mount(formRouterWithProvider('/books'));
        expect(wrapper.find(Books).length).toEqual(1);
        expect(wrapper.find(Home).length).toEqual(0);
    });

    it('renders the BookForm', () => {
        const wrapper = mount(formRouterWithProvider('/books/create'));
        expect(wrapper.find(BookForm).length).toEqual(1);
        expect(wrapper.find(Books).length).toEqual(0);
    });

    it('Header text is Blog', () => {
        const wrapper = mount(app);
        const headerWrapper = wrapper.find(Header);

        expect(headerWrapper.find('.header-logo').first().text()).toEqual('Blog');
    });

    it('renders the Footer', () => {
        const wrapper = mount(app);
        expect(wrapper.find(Footer).length).toEqual(1);
    });

    // it('should change the state after change the input value', () => {
    //     const newValue = 'testing component';
    //     const wrapper = mount(formRouterWithProvider('/books/create', <BookForm match={{params: {}}} location={{state: {}}}/>));
    //
    //     expect(wrapper.find(Field).length).toEqual(5);
    //
    //     // const field = wrapper.find(Field).first()
    //
    //     // const input = wrapper.find('#title');
    //     // field.simulate('change', { target: { value: newValue }});
    //     // expect(field.value).toEqual(newValue);
    // });
})
