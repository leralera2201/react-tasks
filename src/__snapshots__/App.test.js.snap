// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<Fragment>
  <Header />
  <ToastContainer
    autoClose={5000}
    closeButton={[Function]}
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position='top-right'
    role='alert'
    rtl={false}
    transition={[Function]}
  />
  <div
    className='wrapper'
  >
    <Switch>
      <Route
        component={[Function]}
        exact={true}
        path='/'
      />
      <Route
        component={
          Object {
            '$$typeof': Symbol(react.memo),
            'WrappedComponent': [Function],
            'compare': null,
            'type': [Function],
          }
        }
        exact={true}
        path='/books'
      />
      <Route
        component={
          Object {
            '$$typeof': Symbol(react.memo),
            'WrappedComponent': [Function],
            'compare': null,
            'type': [Function],
          }
        }
        exact={true}
        path='/books/create'
      />
      <Route
        component={
          Object {
            '$$typeof': Symbol(react.memo),
            'WrappedComponent': [Function],
            'compare': null,
            'type': [Function],
          }
        }
        exact={true}
        path='/books/:id'
      />
      <Route
        component={
          Object {
            '$$typeof': Symbol(react.memo),
            'WrappedComponent': [Function],
            'compare': null,
            'type': [Function],
          }
        }
        exact={true}
        path='/books/:id/update'
      />
    </Switch>
  </div>
  <Footer />
</Fragment>
`;
